src/Features/Core/Portable/TodoComments/AbstractTodoCommentService.cs

```diff
@@ -42,9 +55,32 @@ internal abstract class AbstractTodoCommentService : ITodoCommentService
         private async Task<IList<TodoComment>> GetTodoCommentsInRemoteHostAsync(
             RemoteHostClient client, Document document, ImmutableArray<TodoCommentDescriptor> commentDescriptors, CancellationToken cancellationToken)
         {
-            return await client.RunCodeAnalysisServiceOnRemoteHostAsync<IList<TodoComment>>(
-                document.Project.Solution, nameof(IRemoteTodoCommentService.GetTodoCommentsAsync),
-                new object[] { document.Id, commentDescriptors }, cancellationToken).ConfigureAwait(false);
+            var keepAliveSession = await GetKeepAliveSessionAsync(client).ConfigureAwait(false);
+
+            var (success, result) = await keepAliveSession.TryInvokeAsync<IList<TodoComment>>(
+                nameof(IRemoteTodoCommentService.GetTodoCommentsAsync),
+                document.Project.Solution,
+                new object[] { document.Id, commentDescriptors }).ConfigureAwait(false);
+
+            return success ? result : SpecializedCollections.EmptyList<TodoComment>();
+        }
+
+        private async Task<KeepAliveSession> GetKeepAliveSessionAsync(RemoteHostClient client)
+        {
+            if (_session != null)
+            {
+                return _session;
+            }
+
+            var session = await client.TryCreateServiceKeepAliveSessionAsync(WellKnownServiceHubServices.CodeAnalysisService, CancellationToken.None).ConfigureAwait(false);
+            var oldSession = Interlocked.CompareExchange(ref _session, session, null);
+            if (oldSession != null)
+            {
+                // there was race, kill the lost session
```

Comment:
could we just have a SemaphoreSlim here?  i really hate having to try to figure out if this sort of code is correct, when a nice simple gate would be totally fine and would be completely clear.  we have cancellation tokens and async waiting, so it's totally fine to use that pattern.